src/Analysis/Elaboration.hs

```diff
@@ -1,87 +1,99 @@
+  | InfiniteType Name (Type (PartialType Error))
   deriving (Eq, Ord, Read, Show)
 
-type Elab = StateT (Subst (Free Type Error)) (ReaderT (Env (Fix Type)) (Fresh Name))
+type Elab = StateT (Subst (PartialType Error)) (ReaderT (Env Name) (Fresh Name))
 
-runElab :: Elab a -> (a, Subst (Free Type Error))
-runElab = fst . flip runFresh (Name 0) . flip runReaderT mempty . flip runStateT mempty
-
-substElaborated :: Cofree Expr (Free Type Error) -> Subst (Free Type Error) -> Cofree Expr (Free Type Error)
-substElaborated = cata (\ (ty F.:< expr) subst -> substitute subst ty :< (($ subst) <$> expr))
+type PartialElabTerm = Cofree Expr (PartialType Error)
 
+runElab :: Elab a -> (a, Subst (PartialType Error))
+runElab = fst . flip runFresh (Name 0) . flip runReaderT mempty . flip runStateT mempty
 
-elaborate :: Fix Expr -> Elab (Cofree Expr (Free Type Error))
+elaborate :: Term -> Elab PartialElabTerm
 elaborate (Fix (Abs n b)) = do
   t <- fresh
-  b' <- local (envExtend n (Fix (TVar t))) (elaborate b)
-  pure (Free (Free (TVar t) :-> extract b') :< Abs n b')
+  b' <- local (envExtend n t) (elaborate b)
+  pure ((tvar t .-> extract b') :< Abs n b')
 elaborate (Fix (App f a)) = do
   t <- fresh
-  f' <- elaborate f
   a' <- elaborate a
-  fTy <- unify (extract f') (Free (extract a' :-> Free (TVar t)))
-  pure (fromMaybe (Free (TVar t)) (returnType fTy) :< App f' a')
+  f' <- check f (extract a' .-> tvar t)
+  pure (tvar t :< App f' a')
 elaborate (Fix (Var name)) = do
   env <- ask
-  pure (maybe (Pure (FreeVariable name)) (cata Free) (envLookup name env) :< Var name)
-elaborate (Fix (Lit b)) = pure (Free Bool :< Lit b)
+  pure (maybe (Pure (FreeVariable name)) tvar (envLookup name env) :< Var name)
+elaborate (Fix (Lit b)) = pure (bool :< Lit b)
 elaborate (Fix (If c t e)) = do
-  c' <- elaborate c
+  c' <- check c bool
   t' <- elaborate t
   e' <- elaborate e
   result <- unify (extract t') (extract e')
   pure (result :< If c' t' e')
 elaborate (Fix (Pair fst snd)) = do
   fst' <- elaborate fst
   snd' <- elaborate snd
-  pure (Free (extract fst' :* extract snd') :< Pair fst' snd')
+  pure (extract fst' .* extract snd' :< Pair fst' snd')
 elaborate (Fix (Fst pair)) = do
   t1 <- fresh
   t2 <- fresh
-  pair' <- elaborate pair
-  pairTy <- unify (extract pair') (Free (Free (TVar t1) :* Free (TVar t2)))
-  pure (fromMaybe (Free (TVar t1)) (fstType pairTy) :< Fst pair')
+  pair' <- check pair (tvar t1 .* tvar t2)
+  pure (tvar t1 :< Fst pair')
 elaborate (Fix (Snd pair)) = do
   t1 <- fresh
   t2 <- fresh
-  pair' <- elaborate pair
-  pairTy <- unify (extract pair') (Free (Free (TVar t1) :* Free (TVar t2)))
-  pure (fromMaybe (Free (TVar t2)) (sndType pairTy) :< Snd pair')
+  pair' <- check pair (tvar t1 .* tvar t2)
+  pure (tvar t2 :< Snd pair')
 
+check :: Term -> PartialType Error -> Elab (Cofree Expr (PartialType Error))
+check term ty = do
+  term' <- elaborate term
+  termTy <- unify (extract term') ty
+  pure (termTy :< unwrap term')
```

Comment:
This isn’t really a bidirectional type checker, just a helper to apply unification.